### PR TITLE
[WIP] Grant 'superuser' to the members of  Postgres admin teams

### PR DESCRIPTION
**Pull request by zerg-junior**
Initial version, do not merge.

### CODE REVIEW

File: pkg/cluster/cluster.go [hunk 1]
```diff
@@ -760,6 +763,17 @@ func (c *Cluster) initHumanUsers() error {
 	return nil
 }
 
+func (c *Cluster) initHumanUsers() error {
+
+	for _, postgresAdminTeam := range c.OpConfig.PostgresAdminTeams {
+		c.initTeamMembers(postgresAdminTeam, true /*isPostgresAdminTeam*/)
```
Comment (golangci) [inline pkg/cluster/cluster.go]: Error return value of `c.initTeamMembers` is not checked

File: pkg/cluster/cluster.go [hunk 2]
```diff
@@ -760,6 +763,17 @@ func (c *Cluster) initHumanUsers() error {
...
+	}
+
+	c.initTeamMembers(c.Spec.TeamID, false /*isPostgresAdminTeam*/)
```
Comment (golangci) [inline pkg/cluster/cluster.go]: Error return value of `c.initTeamMembers` is not checked

File: docs/administrator.md [hunk 3]
```diff
@@ -208,3 +208,13 @@ generated from the current cluster manifest. There are two types of scans: a
...
+
+## Postgres roles supported by the operator
+
+The operator is capable of maintaining roles of multiple kinds within a Postgres database cluster:
+
+1. **System roles** are roles necessary for the proper work of Postgres itself such as a replication role or the initial superuser role. Operator delegates creating such roles to Patroni and only establishes relevant secrets.
+
+2. **Infrastructure roles** are roles for processes originating from external systems, e.g. monitoring robots. The operator creates such roles assuming k8s secrets with their credentials exist beforehand.
+
+3. **Human users**. Human users can originate from either the Teams API or a Postgres manifest. Operator differentiates between product teams that may be granted admin rights to maintain there own Postgres cluster, and the Postgres administrator teams that get superuser access to all PG databases running in a k8s cluster for the purposes of maintaining and troubleshooting.
```
Comment (alexeyklyukin) [inline docs/administrator.md]: There are also per-cluster robot users.

File: pkg/cluster/cluster.go (hunk 2 above)
Comment (alexeyklyukin) [inline pkg/cluster/cluster.go]: What if your team is already listed in both the admin and the owner team, would resolveTeamMembers handle this correctly?

File: pkg/cluster/util.go [hunk 4]
```diff
@@ -225,9 +227,9 @@ func (c *Cluster) getTeamMembers() ([]string, error) {
 		return []string{}, nil
 	}
 
-	teamInfo, err := c.teamsAPIClient.TeamInfo(c.Spec.TeamID, token)
+	teamInfo, err := c.teamsAPIClient.TeamInfo(teamID, token)
 	if err != nil {
-		c.logger.Warnf("could not get team info, returning empty list of team members: %v", err)
+		c.logger.Warnf("could not get team info for team %v, returning empty list of team members: %v", teamID, err)
```
Comment (alexeyklyukin) [inline pkg/cluster/util.go]: please, use %q instead of %v for simple string variables.

File: docs/reference/operator_parameters.md [hunk 5]
```diff
@@ -377,6 +377,9 @@ key.
   List of roles that cannot be overwritten by an application, team or
   infrastructure role. The default is `admin`.
 
+* **postgres_admin_teams**
```
Comment (Jan-M) [inline docs/reference/operator_parameters.md]: This is confusing. We give "admin" to owning teams and call the superusers admins. Let's call it by what is is. postgres_superuser_teams?

File: docs/administrator.md (hunk 3 above)
Comment (zerg-junior) [reply docs/administrator.md]: By robot users here do we mean for instance roles created for applications connections ?

File: pkg/cluster/cluster.go (hunk 1 above)
Comment (Jan-M) [reply pkg/cluster/cluster.go]: Imho no need for this comment in the function call.

File: pkg/util/config/config.go [hunk 6]
```diff
@@ -103,6 +103,7 @@ type Config struct {
 	TeamAPIRoleConfiguration map[string]string `name:"team_api_role_configuration" default:"log_statement:all"`
 	PodTerminateGracePeriod  time.Duration     `name:"pod_terminate_grace_period" default:"5m"`
 	ProtectedRoles           []string          `name:"protected_role_names" default:"admin"`
+	PostgresAdminTeams       []string          `name:"postgres_admin_teams" default:"postgres_admin"`
```
Comment (Jan-M) [inline pkg/util/config/config.go]: Please make this an empty default, don't want people to create a team called postgres_admin and suddenly get super power.

File: pkg/util/config/config.go [hunk 7]
```diff
@@ -103,7 +103,7 @@ type Config struct {
 	TeamAPIRoleConfiguration map[string]string `name:"team_api_role_configuration" default:"log_statement:all"`
 	PodTerminateGracePeriod  time.Duration     `name:"pod_terminate_grace_period" default:"5m"`
 	ProtectedRoles           []string          `name:"protected_role_names" default:"admin"`
-	PostgresAdminTeams       []string          `name:"postgres_admin_teams" default:"postgres_admin"`
+	PostgresSuperuserTeams   []string          `name:"postgres_superuser_teams" default:"postgres_superusers"`
```
Comment (Jan-M) [inline pkg/util/config/config.go]: Still with a default ;)

File: pkg/cluster/cluster.go [hunk 8]
```diff
@@ -760,6 +763,30 @@ func (c *Cluster) initHumanUsers() error {
...
+func (c *Cluster) initHumanUsers() error {
+
+	for _, postgresSuperuserTeam := range c.OpConfig.PostgresSuperuserTeams {
+		err := c.initTeamMembers(postgresSuperuserTeam, true /*isPostgresSuperuserTeam*/)
+		if err != nil {
+			return fmt.Errorf("Cannot create a team %q of Postgres superusers: %v", postgresSuperuserTeam, err)
+		}
+	}
+
+	for _, postgresSuperuserTeam := range c.OpConfig.PostgresSuperuserTeams {
```
Comment (Jan-M) [inline pkg/cluster/cluster.go]: is there no "in" operator? for String array?

File: pkg/cluster/cluster.go [hunk 9]
```diff
@@ -760,6 +763,30 @@ func (c *Cluster) initHumanUsers() error {
...
+		if postgresSuperuserTeam == c.Spec.TeamID {
+			c.logger.Infof("Team %q owning the cluster is also a team of superusers. Created superuser roles for its members instead of admin roles.", c.Spec.TeamID)
+			return nil
+		}
+	}
+
+	err := c.initTeamMembers(c.Spec.TeamID, false /*isPostgresSuperuserTeam*/)
```
Comment (Jan-M) [inline pkg/cluster/cluster.go]: Seems this can be skipped if superuser team contains team already? (hinted by Oleksii)

Comment (zerg-junior) [reply pkg/cluster/cluster.go]: it is skipped in the lines ahead

File: pkg/cluster/cluster.go (hunk 8 above)
Comment (zerg-junior) [reply pkg/cluster/cluster.go]: [apparently no due to the lack of generics in the language](https://stackoverflow.com/questions/10485743/contains-method-for-a-slice)

Comment (erthalion) [reply pkg/cluster/cluster.go]: @zerg-junior but why this part can't be merged with the previous loop?

File: pkg/cluster/cluster.go [hunk 10]
```diff
@@ -760,6 +763,30 @@ func (c *Cluster) initHumanUsers() error {
 	return nil
 }
 
+func (c *Cluster) initHumanUsers() error {
+
+	for _, postgresSuperuserTeam := range c.OpConfig.PostgresSuperuserTeams {
+		err := c.initTeamMembers(postgresSuperuserTeam, true /*isPostgresSuperuserTeam*/)
```
Comment (erthalion) [inline pkg/cluster/cluster.go]: I would say if we need this commentary, it's better to incorporate this information into the argument itself - e.g. replace boolean with enum.

Comment (zerg-junior) [reply pkg/cluster/cluster.go]: this is just  a convenience commentary, imo not worth an enum for just two cases

**Comment by Jan-M** [comment]
:+1:

**Pull request merged**